add_to_app/android_books/app/src/main/java/dev/flutter/example/books/FlutterBookActivity.kt

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package dev.flutter.example.books
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import dev.flutter.example.books.BookApplication.Companion.ENGINE_ID
+import io.flutter.embedding.android.FlutterActivity
+import io.flutter.embedding.engine.FlutterEngine
+import java.util.HashMap
+
+/**
+ * This {@link FlutterActivity} class repackages Kotlin-Dart interop using the Pigeon IPC mechanism.
+ * It repackages Flutter/Dart-side functionalities in standard Android API style, passing
+ * arguments in and out of the activity using 'startActivityForResult' intents and
+ * 'onActivityResult' intents.
+ */
+class FlutterBookActivity: FlutterActivity() {
+    companion object {
+        const val EXTRA_BOOK = "book"
+
+        /**
+         * Static intent factory to start {@link FlutterBookActivity} with the singleton
+         * {@link FlutterEngine} the application started.
+         *
+         * The activity launched from this intent shows the details of the {@link Api.Book}
+         * supplied.
+         */
+        fun withBook(context: Context, book: Api.Book): Intent {
+            // In a more realistic app, there should be some dependency injection mechanism to
+            // determine which engine to use.
+            return CachedEngineBookIntentBuilder(ENGINE_ID)
+                .build(context)
+                .putExtra(
+                    // The Pigeon data class is useful not only between Kotlin/Java and Dart
+                    // but also within Kotlin/Java where activities must communicate with
+                    // each other via serializable data. The Pigeon data class is a
+                    // serializable class by definition.
+                    EXTRA_BOOK,
+                    // TODO(gaaclarke): the Pigeon generated data class should just implement
+                    // Serializable so we won't need 'toMap()' here
+                    // https://github.com/flutter/flutter/issues/58909
+                    book.toMap()
```

Comment:
On other platforms toMap and fromMap are private.  I should make them private or visible for testing only in java too, not package visible.  We don't want people to get reliant on them.

Reply:
+1. We shouldn't let people have access to this. Though we should make sure it implements Serializable first so that the data classes can be used in people's existing projects (needed to share data classes between activities) in addition to just the interop.

Reply:
I don't think so, once we provide a serialization format we have to be backward compatible.  By hiding the actual serialization behind functions we can change it at any point.

Reply:
Discussed offline. This is a widely used Android pattern.